src/wasm-interpreter.h

```diff
@@ -3494,6 +3499,83 @@ class ModuleRunnerBase : public ExpressionRunner<SubType> {
     }
     return {};
   }
+  Flow visitArrayNewSeg(ArrayNewSeg* curr) {
+    NOTE_ENTER("ArrayNewSeg");
+    auto offsetFlow = self()->visit(curr->offset);
+    if (offsetFlow.breaking()) {
+      return offsetFlow;
+    }
+    auto sizeFlow = self()->visit(curr->size);
+    if (sizeFlow.breaking()) {
+      return sizeFlow;
+    }
+
+    auto offset = offsetFlow.getSingleValue().geti32();
+    auto size = sizeFlow.getSingleValue().geti32();
+
+    auto heapType = curr->type.getHeapType();
+    const auto& element = heapType.getArray().element;
+    auto elemType = heapType.getArray().element.type;
+
+    Literals contents;
+    contents.reserve(size);
+
+    switch (curr->op) {
+      case NewData: {
+        assert(curr->segment < wasm.dataSegments.size());
+        assert(elemType.isNumber());
+        const auto& seg = *wasm.dataSegments[curr->segment];
+        auto elemBytes =
+          element.packedType == Field::i8
+            ? 1
+            : element.packedType == Field::i16 ? 2 : elemType.getByteSize();
+
+        auto end = (uint64_t)offset + size * elemBytes;
+        if ((size != 0ull && droppedSegments.count(curr->segment)) ||
+            end > seg.data.size()) {
+          trap("out of bounds segment access in array.new_data");
+        }
+        for (Index i = offset; i < end; i += elemBytes) {
+          auto addr = (void*)&seg.data[i];
+          switch (element.packedType) {
+            case Field::not_packed:
+              contents.push_back(Literal::makeFromMemory(addr, elemType));
+              break;
+            case Field::i8: {
+              uint8_t val;
+              memcpy(&val, addr, 1);
+              contents.push_back(Literal(int32_t(val)));
+              break;
+            }
+            case Field::i16: {
+              uint16_t val;
+              memcpy(&val, addr, 2);
+              contents.push_back(Literal(int32_t(val)));
+              break;
+            }
```

Comment:
Maybe these memcpys could also be in `Literal::makeFromMemory`? If that were in `literal.cpp` (as separately suggested above) then we could just forward-declare `Field` I think and it would work.